retrieve/retrieve.go

```diff
@@ -0,0 +1,1091 @@
+		c.rep.Publish(
+			rep.NewRetrievalEventFailure(rep.RetrievalPhase, rootCid, peerID, address.Undef, err.Error()))
+		return nil, err
+	} else if !has {
+		msg := "data transfer failed: unconfirmed block transfer"
+		c.rep.Publish(
+			rep.NewRetrievalEventFailure(rep.RetrievalPhase, rootCid, peerID, address.Undef, msg))
+		return nil, errors.New(msg)
+	}
+
+	// Compile the retrieval stats
+
+	state, err := c.dataTransfer.ChannelState(ctx, chanid)
+	if err != nil {
+		err = fmt.Errorf("could not get channel state: %w", err)
+		c.rep.Publish(
+			rep.NewRetrievalEventFailure(rep.RetrievalPhase, rootCid, peerID, address.Undef, err.Error()))
+		return nil, err
+	}
+
+	duration := time.Since(startTime)
+	speed := uint64(float64(state.Received()) / duration.Seconds())
+
+	// Otherwise publish a retrieval event success
+	c.rep.Publish(rep.NewRetrievalEventSuccess(rep.RetrievalPhase, rootCid, peerID, address.Undef, state.Received(), state.ReceivedCidsTotal(), duration, totalPayment))
+
+	return &RetrievalStats{
+		Peer:         state.OtherPeer(),
+		Size:         state.Received(),
+		Duration:     duration,
+		AverageSpeed: speed,
+		TotalPayment: totalPayment,
+		NumPayments:  int(nonce),
+		AskPrice:     proposal.PricePerByte,
+	}, nil
+}
+
+func (fc *Client) SubscribeToRetrievalEvents(subscriber rep.RetrievalSubscriber) {
+	fc.rep.Subscribe(subscriber)
+}
+
+// Implement RetrievalSubscriber
+func (fc *Client) OnRetrievalEvent(event rep.RetrievalEvent) {
+	kv := make([]interface{}, 0)
+	logadd := func(kva ...interface{}) {
+		if len(kva)%2 != 0 {
+			panic("bad number of key/value arguments")
+		}
+		for i := 0; i < len(kva); i += 2 {
+			key, ok := kva[i].(string)
+			if !ok {
+				panic("expected string key")
+			}
+			kv = append(kv, key, kva[i+1])
+		}
+	}
+	logadd("code", event.Code(),
+		"phase", event.Phase(),
+		"payloadCid", event.PayloadCid(),
+		"storageProviderId", event.StorageProviderId(),
+		"storageProviderAddr", event.StorageProviderAddr())
+	switch tevent := event.(type) {
+	case rep.RetrievalEventQueryAsk:
+		logadd("queryResponse:Status", tevent.QueryResponse().Status,
+			"queryResponse:PieceCIDFound", tevent.QueryResponse().PieceCIDFound,
+			"queryResponse:Size", tevent.QueryResponse().Size,
+			"queryResponse:PaymentAddress", tevent.QueryResponse().PaymentAddress,
+			"queryResponse:MinPricePerByte", tevent.QueryResponse().MinPricePerByte,
+			"queryResponse:MaxPaymentInterval", tevent.QueryResponse().MaxPaymentInterval,
+			"queryResponse:MaxPaymentIntervalIncrease", tevent.QueryResponse().MaxPaymentIntervalIncrease,
+			"queryResponse:Message", tevent.QueryResponse().Message,
+			"queryResponse:UnsealPrice", tevent.QueryResponse().UnsealPrice)
+	case rep.RetrievalEventFailure:
+		logadd("errorMessage", tevent.ErrorMessage())
+	case rep.RetrievalEventSuccess:
+		logadd("receivedSize", tevent.ReceivedSize())
+	}
+	log.Debugw("retrieval-event", kv...)
+}
+
+func ChannelIDFromString(id string) (*datatransfer.ChannelID, error) {
```

Comment:
Unused

Reply:
Wonder why the linter is not catching this? It does catch "some" unused methods.